### PR TITLE
Assettransfer

### PR DESCRIPTION
**Pull request by Suja-Muthukrishnan**
TransferAsset

### CODE REVIEW

File: lambda/api/provision-device/index.js [hunk 1]
```diff
@@ -5,32 +5,88 @@ let tclAws = require('tcl-aws');
...
-
+    let promises = [];
+    let okCount = 0;
+    let totalResourcesCount = 0;
     return validateData(event)
-        .then(data => {
+        .then((data) => {
             console.log(`Validation result:`, JSON.stringify(data));
-            return vehicleRename(data.vehicleId, data.requestBody.vehicleNewName)
+            if (data.requestBody.vehicleNewName) {
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: I think you are making the logic harder on yourself than you need to by adding the conditional statements here. If I'm not mistaken, the validateData would not allow you to get to this point if the request didn't have enough information to perform the rename and transfer operations... both are required, right?

Since that's the case you could simplify this section by doing the following:
```
    return validateData(event)
        .then((data) => {
            console.log(`Validation result:`, JSON.stringify(data));
            promises.push(ProcessVehicleRename(data));
            promises.push(ProcessVehicleTransfer(data));
            return Promise.all(promises);
        })
        .then((res) => {
            if (res[0].statusCode === 200 && res[1].statusCode === 200) {
                return done(null, res, null);
            } else {
                console.log('One or more of the promises failed. Unhandled exception processing requests so we can retry.', res);
                if (res[0].statusCode === 200 || res[1].statusCode === 200) {
                    return done(null, res, MULTI_STATUS);
                } else {
                    return done(null, res, 500);
                }
            }
        })
```

If you don't refactor the code in this way, then you have to address my comments below about how you handle the cases where `okCount` is zero. See below.

File: lambda/api/provision-device/index.js [hunk 2]
```diff
@@ -5,32 +5,88 @@ let tclAws = require('tcl-aws');
...
+                console.log('One of the Promise got Failed, Unhandled exception processing requests. so we can retry.', res);
+                return done(null, res, MULTI_STATUS)
+            }
+            if (!okCount) {
+                if (totalResourcesCount == 1) {
+                    return done(res, null, res[0].statusCode);
+                }
+                else {
+                    console.log('PromiseFailed, Unhandled exception processing requests. so we can retry.', res);
+                    return done(res, null, MULTI_STATUS);
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: If `okCount` is zero and `totalResourcesCount` is also zero, I don't think that's a `207`, is it? That's a complete failure in that case. In reality, any scenario in which `totalResourcesCount` is < 1 is a 4xx error because somehow the request/client hasn't provided enough information to properly perform this operation.

**Comment by ThomasSunderland** [comment]
Does this PR have 13 commits??? You should really squash the commits before opening a PR. Also it appears it has conflicts with the master branch and cannot be merged as-is. This should probably also be resolved before opening the PR.

File: lambda/api/provision-device/index.js [hunk 3]
```diff
@@ -5,32 +5,88 @@ let tclAws = require('tcl-aws');
...
+            console.log(`PromiseFailed: Unhandled exception processing VehicleTransfer requests. Failing so we can retry.`, err);
+            return Promise.resolve(err);
+        });
+}
 
 function validateData(event) {
     let deviceId = undefined;
     let tempBearerToken = undefined;
-    let targetOrgId = undefined;
+    let temptargetOrgId = undefined;
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: Please fix the casing on these two variables...
`tempTargetOrgId`
`tempAssetId`

File: lambda/api/provision-device/index.js [hunk 4]
```diff
@@ -64,14 +120,46 @@ function validateData(event) {
...
+    temptargetOrgId = body.vehicleTargetOrgId;
+    tempassetId = body.assetId
+    let isInvalidName = true;
+    let isInvalidOrgId = true;
+    let isInvalidAssetId = true;
+
+    isInvalidName = !body.vehicleNewName;
+    isInvalidOrgId = (!temptargetOrgId || isNaN(temptargetOrgId) || temptargetOrgId <= 0);
+    isInvalidAssetId = (!tempassetId || isNaN(tempassetId) || tempassetId <= 0);
+    if (body.hasOwnProperty('vehicleNewName') && (isInvalidName)) {
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: I don't understand the explicit inclusion of `hasOwnProperty`... if `!body.vehicleName` is true, then there isn't a vehicle name... right? Is there some edge case I'm missing that `hasOwnProperty` catches?

File: lambda/api/provision-device/index.js [hunk 5]
```diff
@@ -64,14 +120,46 @@ function validateData(event) {
...
         return Promise.reject({
             "statusCode": 400,
-            "message": "Bad request. Missing or invalid parameters, vehicle new name is invalid"
+            "message": "Bad request. Missing or invalid parameters, vehicleNewName is invalid"
+        });
+
+    }
+    if (body.hasOwnProperty('vehicleTargetOrgId')) {
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: Why not simply... 
```
    if (isInvalidName && isInvalidOrgId) {
        return Promise.reject({
            "statusCode": 400,
            "message": "Bad request. Missing or invalid parameters, vehicleNewName/vehicleTargetOrgId is invalid"
        });
    } else {
        if (isInvalidOrgId) {
            return Promise.reject({
                "statusCode": 400,
                "message": "Bad request. Missing or invalid parameters, vehicleTargetOrgId is invalid"
            });
        }
        if (isInvalidAssetId) {
            return Promise.reject({
                "statusCode": 400,
                "message": "Bad request. Missing or invalid parameters, assetId is invalid"
            });
        }
    }
```
What is the property check doing? I am finding it hard to follow the logic in this section.

File: lambda/api/provision-device/index.js [hunk 6]
```diff
@@ -146,24 +234,86 @@ function vehicleRename(vehicleId, vehicleNewName) {
...
+        .then((data) => {
+            console.log(`vehicleTransfer returned data:`, data);
+            if (data && data.statusCode && data.statusCode == '200' && data.statusMessage) {
+                return Promise.resolve({
+                    "statusCode": data.statusCode,
+                    "message": "vehicle Transfer successfully done"
+                });
+            } else {
+                return Promise.reject({
+                    "statusCode": data.statusCode,
```
Comment (sforcier) [inline lambda/api/provision-device/index.js]: If the response has a status code of `200` but the `statusMessage` is null or empty, then you're going to reject it, but return a status code of `200`.

**Comment by sforcier** [comment]
> Does this PR have 13 commits??? You should really squash the commits before opening a PR. Also it appears it has conflicts with the master branch and cannot be merged as-is. This should probably also be resolved before opening the PR.

I definitely prefer things squished prior to PR, but in this case I believe this was opened with only a few commits, the main one and a couple "touch ups". The remaining commits are related to getting tests running, with the exception of the last one, which seems like a functional change. I do question why more functionality is being added here instead of having closed this PR off after the comments were addressed and tests were green. I'm sure there's a reason, but this PR has been around a while and it's going to start to get more and more difficult to merge if an emphasis isn't placed on getting it merged.

**Comment by Suja-Muthukrishnan** [comment]
> > Does this PR have 13 commits??? You should really squash the commits before opening a PR. Also it appears it has conflicts with the master branch and cannot be merged as-is. This should probably also be resolved before opening the PR.
> 
> I definitely prefer things squished prior to PR, but in this case I believe this was opened with only a few commits, the main one and a couple "touch ups". The remaining commits are related to getting tests running, with the exception of the last one, which seems like a functional change. I do question why more functionality is being added here instead of having closed this PR off after the comments were addressed and tests were green. I'm sure there's a reason, but this PR has been around a while and it's going to start to get more and more difficult to merge if an emphasis isn't placed on getting it merged.

**Comment by Suja-Muthukrishnan** [comment]
> Does this PR have 13 commits??? You should really squash the commits before opening a PR. Also it appears it has conflicts with the master branch and cannot be merged as-is. This should probably also be resolved before opening the PR.

I agree. I did it wrongly. i will correct the same.

**Comment by Suja-Muthukrishnan** [comment]
> > > Does this PR have 13 commits??? You should really squash the commits before opening a PR. Also it appears it has conflicts with the master branch and cannot be merged as-is. This should probably also be resolved before opening the PR.
> > 
> > 
> > I definitely prefer things squished prior to PR, but in this case I believe this was opened with only a few commits, the main one and a couple "touch ups". The remaining commits are related to getting tests running, with the exception of the last one, which seems like a functional change. I do question why more functionality is being added here instead of having closed this PR off after the comments were addressed and tests were green. I'm sure there's a reason, but this PR has been around a while and it's going to start to get more and more difficult to merge if an emphasis isn't placed on getting it merged.

Mistakenly created pull request with previous commits. Sorry for that. I will correct it.

File: docs/swagger.yaml [hunk 7]
```diff
@@ -35,7 +35,7 @@ paths:
           in: body
       responses:
         '200':
-          description: Success. The body will contain "OK" as a response message.
+          description: Success. The body will contain Sucess response message object.
```
[the remark this answers is not in the document]
Comment (Suja-Muthukrishnan) [reply docs/swagger.yaml]: Corrected.

File: lambda/api/provision-device/index.js (hunk 4 above)
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: We can do  rename alone or transfer  alone or both(rename and transfer).  To identify whether we are going to do rename, checking  with hasOwnProperty('vehicleNewName').  if we are not going to do rename, TFMA will not pass 'vehicleNewName' property.

File: lambda/api/provision-device/index.js [hunk 8]
```diff
@@ -5,32 +5,88 @@ let tclAws = require('tcl-aws');
...
+    let totalResourcesCount = 0;
     return validateData(event)
-        .then(data => {
+        .then((data) => {
             console.log(`Validation result:`, JSON.stringify(data));
-            return vehicleRename(data.vehicleId, data.requestBody.vehicleNewName)
+            if (data.requestBody.vehicleNewName) {
+                promises.push(ProcessVehicleRename(data));
+            }
+            if (data.requestBody.vehicleTargetOrgId && data.requestBody.assetId) {
```
[the remark this answers is not in the document]
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: It is the device's id only.  Will continue with deviceId. not going to rename from the validation function.

File: lambda/api/provision-device/index.js (hunk 3 above)
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: Done.

File: lambda/api/provision-device/index.js (hunk 5 above)
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: We can do rename alone or transfer alone or both(rename or transfer).  To identify whether we going to do transfer checking with  body.hasOwnProperty('vehicleTargetOrgId').

File: lambda/api/provision-device/index.js (hunk 1 above)
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: This provisioning service method doesn't always call all flow (rename, transfer and profile change). There may be a case where only the rename call or transfer call or profile change call has to be done. Because of that if only the respective JSON parameter is passed, the flow will be triggered.

File: lambda/api/provision-device/index.js [hunk 9]
```diff
@@ -5,32 +5,88 @@ let tclAws = require('tcl-aws');
...
+                promises.push(ProcessVehicleTransfer(data));
+            }
+            return Promise.all(promises);
         })
-        .then(data => {
-            console.log('Return data:', JSON.stringify(data));
-            return done(null, data, null)
+        .then((res) => {
+            totalResourcesCount = promises.length;
```
[the remark this answers is not in the document]
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: 1) Changed to totalResponseCount.  2) Removed declaration from line 16 and added to line 30 now.

File: lambda/api/provision-device/index.js (hunk 6 above)
Comment (Suja-Muthukrishnan) [reply lambda/api/provision-device/index.js]: Removed  data.statusMessage condition check.
